include/domain_bridge/domain_bridge.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2021, Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DOMAIN_BRIDGE__DOMAIN_BRIDGE_HPP_
+#define DOMAIN_BRIDGE__DOMAIN_BRIDGE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "domain_bridge/visibility_control.hpp"
+
+namespace domain_bridge
+{
+
+class DomainBridgeImpl;
+
+/// Bridge ROS entities across different domains.
+class DomainBridge
+{
+public:
+  /// Constructor.
+  DOMAIN_BRIDGE_PUBLIC
+  DomainBridge();
+
+  /// Destructor.
+  DOMAIN_BRIDGE_PUBLIC
+  ~DomainBridge();
+
+  /// Start the bridge.
+  /**
+   * Start forwarding ROS traffic between domains.
+   * Only topics registered with \ref bridge_topic() will have their messages forwarded.
+   *
+   * This method blocks indefinitely.
+   */
+  DOMAIN_BRIDGE_PUBLIC
+  void spin();
```

Comment:
Makes sense. Are there any examples or docs on recommended patterns for interacting the the executor?

Reply:
> Makes sense. Are there any examples or docs on recommended patterns for interacting the the executor?

Not that I know of, but either having a getter of the node (or the node base interface) or having an add_to_executor method sound like the most general ways of achieving it.

Reply:
New nodes are created every time a new topic is bridged, so it makes more sense to have an add_to_executor method. I've add this and a new TopicBridgeOptions class for passing around options like the callback group (59cb440)

Regarding the callback group, I'm confused about it's relationship with nodes, subscriptions, and executors. It looks like there's several ways that callback groups can be created and passed around :confused: We can't pass it to the subscription object since it has type `rclcpp::SubscriptionBase`. The only option I see at the moment is to keep a list of callback groups and add them to the executor in add_to_executor, but I don't see what this achieves if they aren't associated with the subscriptions. I've left a TODO in the code for now: 

https://github.com/ros2/domain_bridge/blob/59cb44047d4b6cbe95994c3f0769db3afde6560d/src/domain_bridge/domain_bridge.cpp#L151